include/deal.II/base/big_mpi_compat.h

```diff
@@ -0,0 +1,331 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2022 - 2022 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+/**
+ *
+ * BigMPICompat - a tiny MPI 4.x compatibility library
+ *
+ * Release under MIT at https://github.com/tjhei/BigMPICompat
+ */
+#ifndef BIG_MPI_COMPAT_H
+#define BIG_MPI_COMPAT_H
+
+#include <deal.II/base/config.h>
+
+#ifdef DEAL_II_WITH_MPI
+
+#  include <mpi.h>
+#  ifndef MPI_VERSION
+#    error "Your MPI implementation does not define MPI_VERSION!"
+#  endif
+
+#  include <iostream>
+
+
+#  if MPI_VERSION < 3
+#    error "BigMPICompat requires at least MPI 3.0"
+#  endif
+
+/**
+ * This namespace contains symbols related to the BigMPICompat library
+ * to support large MPI routines on MPI implementations that implement
+ * version 3.x of the standard.
+ */
+namespace BigMPICompat
+{
+  static constexpr MPI_Count mpi_max_count = (1ULL << 31);
+
+} // namespace BigMPICompat
+
+#  if MPI_VERSION >= 4
+
+#  else
+
+/**
+ * Create a contiguous type of (possibly large) @p count.
+ *
+ * See the MPI 4.x standard for details.
+ */
+inline int
+MPI_Type_contiguous_c(MPI_Count     count,
+                      MPI_Datatype  oldtype,
+                      MPI_Datatype *newtype)
+{
+  if (count <= BigMPICompat::mpi_max_count)
+    return MPI_Type_contiguous(count, oldtype, newtype);
```

Comment:
Is this guaranteed to work if `sizeof(oldtype)*count` is larger than 2^31? Or does it only work if `oldtype` is `MPI_CHAR`?